// posthog-js/lib/src/metrics-defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveMetricsConfig = resolveMetricsConfig;
var DEFAULT_FLUSH_INTERVAL_MS = 10000;
var DEFAULT_MAX_SERIES_PER_FLUSH = 1000;
/**
 * Resolves the public `metrics` config into the shape core `PostHogMetrics`
 * consumes. Same precedence rule as logs: OTLP keys in `resourceAttributes`
 * win over the named config fields.
 */
function resolveMetricsConfig(config) {
    var _a, _b, _c, _d, _e;
    var resourceAttributes = config === null || config === void 0 ? void 0 : config.resourceAttributes;
    return {
        serviceName: (_a = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['service.name']) !== null && _a !== void 0 ? _a : config === null || config === void 0 ? void 0 : config.serviceName,
        serviceVersion: (_b = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['service.version']) !== null && _b !== void 0 ? _b : config === null || config === void 0 ? void 0 : config.serviceVersion,
        environment: (_c = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['deployment.environment']) !== null && _c !== void 0 ? _c : config === null || config === void 0 ? void 0 : config.environment,
        resourceAttributes: resourceAttributes,
        beforeSend: config === null || config === void 0 ? void 0 : config.beforeSend,
        flushIntervalMs: (_d = config === null || config === void 0 ? void 0 : config.flushIntervalMs) !== null && _d !== void 0 ? _d : DEFAULT_FLUSH_INTERVAL_MS,
        maxSeriesPerFlush: (_e = config === null || config === void 0 ? void 0 : config.maxSeriesPerFlush) !== null && _e !== void 0 ? _e : DEFAULT_MAX_SERIES_PER_FLUSH,
    };
}
